homeassistant/components/upnp/sensor.py

```diff
@@ -1,273 +1,243 @@
-        if self._is_overflowed(new_value):
-            self._state = None  # temporarily report nothing
-        else:
-            delta_time = (now - self._last_update_time).seconds
-            delta_value = new_value - self._last_value
-            self._state = delta_value / delta_time
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to sensors events."""
+        self._coordinator.async_add_listener(self.async_write_ha_state)
 
-        self._last_value = new_value
-        self._last_update_time = now
+        async_dispatcher_connect(
+            self.hass, SIGNAL_REMOVE_DEVICE, self._upnp_remove_sensor
+        )
 
+    @callback
+    def _upnp_remove_sensor(self, device) -> None:
+        """Remove sensor."""
+        if self._device != device:
+            # not for us
+            return
 
-class KBytePerSecondUPnPIGDSensor(PerSecondUPnPIGDSensor):
-    """Representation of a KBytes Sent/Received per second sensor."""
+        _LOGGER.debug("Removing sensor: %s", self.unique_id)
+        self.hass.async_create_task(self.async_remove())
 
-    @property
-    def unit(self) -> str:
-        """Get unit we are measuring in."""
-        return DATA_KIBIBYTES
+    async def async_will_remove_from_hass(self) -> None:
+        """When entity will be removed from hass."""
+        self._coordinator.async_remove_listener(self.async_write_ha_state)
 
-    async def _async_fetch_value(self) -> float:
-        """Fetch value from device."""
-        if self._direction == IN:
-            return await self._device.async_get_total_bytes_received()
 
-        return await self._device.async_get_total_bytes_sent()
+class RawUpnpSensor(UpnpSensor):
+    """Representation of a UPnP/IGD sensor."""
 
     @property
     def state(self) -> str:
         """Return the state of the device."""
-        if self._state is None:
-            return None
+        data_name = self._sensor_type["data_name"]
+        value = self._coordinator.data[data_name]["value"]
+        return format(value, "d")
 
-        return format(float(self._state / KIBIBYTE), ".1f")
 
+class DerivedUpnpSensor(UpnpSensor):
+    """Abstract representation of a UNIT Sent/Received per second sensor."""
 
-class PacketsPerSecondUPnPIGDSensor(PerSecondUPnPIGDSensor):
-    """Representation of a Packets Sent/Received per second sensor."""
+    def __init__(self, coordinator, device, sensor_type) -> None:
+        """Initialize sensor."""
+        super().__init__(coordinator, device, sensor_type)
+        self._last_data = None
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return f"{self._device.name} {self._sensor_type['derived_name']}"
 
     @property
-    def unit(self) -> str:
-        """Get unit we are measuring in."""
-        return "packets"
+    def unique_id(self) -> str:
+        """Return an unique ID."""
+        return f"{self._device.name}_{self._device.udn}_{self._sensor_type['derived_name']}"
```

Comment:
Names should never be used for unique IDs. Only udn + sensor type should be enough?